### PR TITLE
Fix SRGConstant layout mismatch causing gpu mem stomp

### PR DESCRIPTION
**Pull request by moudgils**
## What does this PR do?

Atom does not allow partial SRGs to be included in this manner used by the EditorFeedback gem shaders
 - In this case one shader was using another shader's SRGPool due to the fact that both were generating  the same InstanceId which was wrong. For more details please see the attached issue below. 
 - Addressed this by creating one passSRG for all the shaders. This may actually help perf as it will reduce the number of SRGPool instances at the cost of fatter SRG. 
 - Added an assert to catch this issue when a shader is added. 
 - Added a GHI to actually add this feature in future - https://github.com/o3de/o3de/issues/15794

## How was this PR tested?

Tested MPS Editor

## Impact if this code does not go into stabilization

The fix itself is not exactly a crash fix since we are getting lucky that the gpu stomp in this case is not causing a crash but this should go into stab given the severity of the fix. It is likely that a crash may manifest on some combination of hw/driver version out there. 

## Exception Tracking

- [x] 1. Owning SIG requests the exception, by a Sig Representative (Chair/co-chair) adding the PR to the exception requests queue board.
- [x] 2. Owning SIG comments on the PR explaining impact if this code does not go into stabilization.
- [x] 3. Unrelated SIG representative (chair/cochair) approves the exception, adding a comment indicating approval and the SIG they represent.
- [x] 4. UX SIG representative (chair/cochair) approves the exception, adding a comment indicating approval and the SIG they represent.
- [x] 5. Testing SIG representative (chair/cochair) approves the exception, adding a comment indicating approval and the SIG they represent.
- [x] (Must occur at some point before SIG-Release approval) Code has been approved for merging by maintainers
- [x] 6. SIG-Release Release Manager or co-Release Manager approves the exception, adding a comment indicating approval and the SIG they represent.
- [ ] 7. If all approvals are accepted, then the exception PR may be merged
- [ ] 8. Merged PR is verified successful via testing. If Merged PR verification testing failed, consult reps from step 3-5 and discussion occurs on fixing the failure vs backing out change
- [ ] 9. Exception is complete.

### CODE REVIEW

**Review by VickyAtAZ — APPROVED** [review]

File: Gems/AtomLyIntegration/EditorModeFeedback/Assets/Shaders/EditorModeCommon.azsli [hunk 1]
```diff
@@ -44,6 +44,75 @@ partial ShaderResourceGroup PassSrg : SRG_PerPass_WithFallback
...
+
+        V = View position
+        S = Depth transition start
+        E = Depth transition end
+
+        When E is a non-zero value, the binary mask value will be transformed into a value from [m_minDepthTransitionValue, 1.0]
+        according to the fragment's depth position in the transition band (#) spanning from S to E. This is to allow feedback
+        effects to be scaled according to the distance of non-mask fragments to the viewer.
+    */
+
```
Comment (rgba16f) [inline Gems/AtomLyIntegration/EditorModeFeedback/Assets/Shaders/EditorModeCommon.azsli:103]: nit: something to make it visually obvious this is a multiline comment. I'd sugggest either a `*` per line or use `//` on every line.

**Review by rgba16f — APPROVED** [review]

File: Gems/AtomLyIntegration/EditorModeFeedback/Assets/Shaders/EditorModeCommon.azsli (hunk 1 above)
Comment (moudgils) [reply Gems/AtomLyIntegration/EditorModeFeedback/Assets/Shaders/EditorModeCommon.azsli:103]: Given that we need this to go into stab today I wont make this change in this PR but can make it in dev branch later.

**Comment by tonybalandiuk** [comment]
Steps to get this into stabilization are here https://github.com/orgs/o3de/projects/49/views/1?pane=info

File: Gems/AtomLyIntegration/EditorModeFeedback/Assets/Shaders/EditorModeCommon.azsli [hunk 2]
```diff
@@ -76,4 +145,23 @@ partial ShaderResourceGroup PassSrg : SRG_PerPass_WithFallback
...
+        {
+            const float depthTransition = clamp((linearDepth - PassSrg::m_depthTransitionStart) / (PassSrg::m_depthTransitionDuration), PassSrg::m_minDepthTransitionValue, 1.0);
+            return clamp((1.0 - maskValue) * depthTransition, 0.0, 1.0);
+        }
+
+        return 1.0 - maskValue;
+    }
+
+    //! Calculates the blend amount (0.0 = input color, 1.0 = final effect) as scaled by the fragment's non-linear depth according to the depth transition values.
+    float CalculateTransitionBlendAmountFromDepth(const float zDepth, const float maskValue)
```
Comment (lumberyard-employee-dm) [inline Gems/AtomLyIntegration/EditorModeFeedback/Assets/Shaders/EditorModeCommon.azsli:162]: Is the maskValue guaranteed to be between 0 and 1?

**Review by lumberyard-employee-dm — APPROVED** [review]

**Comment by lumberyard-employee-dm** [comment]
As an unrelated SIG representative(@o3de/sig-core chair), I approve of this exception as memory stomps can manifest crashes or corrupte for end users in ways that are non-obvious.

File: Gems/AtomLyIntegration/EditorModeFeedback/Assets/Shaders/EditorModeCommon.azsli (hunk 2 above)
Comment (moudgils) [reply Gems/AtomLyIntegration/EditorModeFeedback/Assets/Shaders/EditorModeCommon.azsli:162]: It seems that way. The mask is provided by red channel of texture which is R8G8_unorm and there is no range modifications happening.

File: Gems/Atom/RPI/Code/Source/RPI.Public/Shader/ShaderResourceGroup.cpp [hunk 3]
```diff
@@ -88,6 +88,8 @@ namespace AZ
 
             m_pool = ShaderResourceGroupPool::FindOrCreate(
                 AZ::Data::Asset<ShaderAsset>(&shaderAsset, AZ::Data::AssetLoadBehavior::PreLoad), supervariantIndex, srgName);
+            AZ_Assert(m_layout->GetHash() == m_pool->GetRHIPool()->GetLayout()->GetHash(), "This can happen if two shaders are including the same partial srg from different .azsl shader files and adding more custom entries to the srg. Recommendation is to just make a bigger SRG that can be shared between the two shaders.");
```
[the remark this answers is not in the document]
Comment (moudgils) [reply Gems/Atom/RPI/Code/Source/RPI.Public/Shader/ShaderResourceGroup.cpp:91]: Will make a mental note and add it as a separate PR to dev branch.

**Comment by AMZN-Dk** [comment]
Reviewed with member of the Testing SIG who is familiar with this area, there is some risk with taking the change however it was evaluated as lower of a risk than not taking the fix. I am approving this from the Testing SIG as co-chair.

**Comment by rainbj** [comment]
Sig-UX-UI approved

**Comment by tonybalandiuk** [comment]
approved from sig-release , please proceed to merge

**Comment by tonybalandiuk** [comment]
please confirm once this has been tested in stabilization

**Comment by moudgils** [comment]
I tested the stab branch on my machine last night after the PR was merged and the Editor was running as expected.

**Pull request merged**